ada_feeding/ada_feeding/behaviors/acquisition/compute_food_frame.py

```diff
@@ -5,56 +5,246 @@
+            ],
+            mask.average_depth,
+        )
+        center = PointStamped()
+        center.header.frame_id = camera_frame
+        center.point.x = center_list[0]
+        center.point.y = center_list[1]
+        center.point.z = center_list[2]
+        center = tf2_ros.TransformRegistration().get(PointStamped)(center, transform)
+
+        # Get angle from mask bounded ellipse
+        # See: https://stackoverflow.com/questions/15956124/minarearect-angles-unsure-
+        #      about-the-angle-returned
+        # Get Mask
+        mask_cv = ros_msg_to_cv2_image(mask.mask)
+        # Threshold and get contours
+        _, mask_thresh = cv.threshold(mask_cv, 127, 255, 0)
+        contours, _ = cv.findContours(mask_thresh, 1, 2)
+        full_contours = np.vstack(contours)
+        rect = cv.minAreaRect(full_contours)
+        points = cv.boxPoints(rect)
+        # Get direction of +X axix in pixel-space
+        # Take longest side
+        if np.linalg.norm(points[0] - points[1]) > np.linalg.norm(
+            points[1] - points[2]
+        ):
+            point1 = points[0]
+            point2 = points[1]
+        else:
+            point1 = points[1]
+            point2 = points[2]
+        # Get vector in camera frame
+        point1 = pyrealsense2.rs2_deproject_pixel_to_point(
+            self.intrinsics, [point1[0], point1[1]], mask.average_depth
+        )
+        point2 = pyrealsense2.rs2_deproject_pixel_to_point(
+            self.intrinsics, [point2[0], point2[1]], mask.average_depth
+        )
+        x_pos = Vector3Stamped()
+        x_pos.header.frame_id = camera_frame
+        x_pos.vector.x = point1[0] - point2[0]
+        x_pos.vector.y = point1[1] - point2[1]
+        x_pos.vector.z = point1[2] - point2[2]
+        # Transform to world frame
+        x_pos = tf2_ros.TransformRegistration().get(Vector3Stamped)(x_pos, transform)
+        # Project to world x-y plane
+        x_pos.vector.z = 0.0
+
+        # Convert to TransformStamped
+        world_to_food_transform.transform.translation.x = center.point.x
+        world_to_food_transform.transform.translation.y = center.point.y
+        world_to_food_transform.transform.translation.z = center.point.z
+
+        x_unit = Vector3Stamped()
+        x_unit.vector.x = 1.0
+        world_to_food_transform.transform.rotation = quat_between_vectors(
+            x_unit.vector, x_pos.vector
+        )
+
+        # Write to blackboard outputs
+        if len(self.blackboard_get("debug_food_frame")) > 0:
+            stb = StaticTransformBroadcaster(self.blackboard_get("ros2_node"))
+            stb.sendTransform(world_to_food_transform)
+            self.blackboard_write("debug_tf_publisher", stb)
+        self.blackboard_write("food_frame", world_to_food_transform)
+        request = AcquisitionSelect.Request()
+        request.food_context = mask
+        self.blackboard_write("action_select_request", request)
+
+        return py_trees.common.Status.SUCCESS
```

Comment:
In practice, once we have the transform how long does it take to do the computations? Because the tree is blocked until the computations are done. If it takes a while, we should instead make it an asynch function that does the computation, and return RUNNING until that asynch function succeeds.

Reply:
Timed: 0.0005796703s (i.e. 580us)
I think we're fine